src/goto-programs/show_goto_functions_xml.cpp

```diff
@@ -0,0 +1,142 @@
+/*******************************************************************\
+
+Module: Goto Program
+
+Author: Thomas Kiley
+
+\*******************************************************************/
+
+#include <goto-programs/show_goto_functions_xml.h>
+#include <iostream>
+#include <sstream>
+
+
+#include <util/xml_expr.h>
+#include <util/cprover_prefix.h>
+#include <util/prefix.h>
+
+#include <langapi/language_util.h>
+
+#include "goto_functions.h"
+#include "goto_model.h"
+
+/*******************************************************************\
+
+Function: show_goto_functions_xmlt::show_goto_functions_xmlt
+
+  Inputs:
+   ns - the namespace to use to resolve names with
+
+ Outputs:
+
+ Purpose: For outputing the GOTO program in a readable xml format.
+
+\*******************************************************************/
+
+show_goto_functions_xmlt::show_goto_functions_xmlt(const namespacet &ns)
+  : ns(ns)
+{}
+
+/*******************************************************************\
+
+Function: show_goto_functions_xmlt::show_goto_functions
+
+  Inputs:
+   goto_functions - the goto functions that make up the program
+
+ Outputs:
+
+ Purpose: Walks through all of the functions in the program and returns
+          an xml object representing all their functions. Produces output
+          like this:
+          <functions>
+            <function name=main, isBodyAvaliable=true, isInternal=false>
+              <instructions>
+                <instruction id=ASSIGN>
+                  <location file="main.c" line="14"/>
+                  <instructionValue>// adsakjsdl
+                     x=y</instructionValue>
+                </instruction>
+              </instructions>
+            </function>
+          </functions>
+
+\*******************************************************************/
+
+xmlt show_goto_functions_xmlt::get_goto_functions(
+  const goto_functionst &goto_functions)
+{
+  xmlt xml_functions=xmlt("functions");
+  for(const auto &function_entry : goto_functions.function_map)
+  {
+    const irep_idt &function_name=function_entry.first;
+    const goto_functionst::goto_functiont &function=function_entry.second;
+
+    xmlt &xml_function=xml_functions.new_element("function");
+    xml_function.set_attribute("name", id2string(function_name));
+    xml_function.set_attribute_bool(
+      "isBodyAvailable", function.body_available());
```

Comment:
Note that we are using camelCase for Json, but underscores for XML (see eg, xml_goto_trace)